enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,369 @@
+page will check the plugin checkbox by default. For other operators, the
+checkbox will not be checked by default. Any plugin can be enabled or disabled
+later through the console cluster settings page.
+
+### Loading Plugins
+
+The console backend will proxy requests to the plugin `Service`. This way the
+`Service` itself does not need to be exposed outside the cluster, and avoids
+any same-origin or certificate errors in the browser loading JavaScript from a
+separate domain. 
+
+On load, console will read the manifest for each plugin. The manifest is a JSON
+document that contains metadata about the plugin and the extensions.
+`extensions` mirrors the API used for static plugins, and the same extension
+points will be available.
+
+```json
+{
+  "$schema": "/path/to/schema/manifest.json",
+  "id": "my-plugin",
+  "version": "1.2.3",
+  "pluginAPI": "4.7.x",
+  "displayName": "My Plugin",
+  "extensions": [
+    {
+      "type": "console.flag",
+      "properties": {
+        "handler": {
+          "$codeRef": "example.testHandler"
+        }
+      }
+    },
+    {
+      "type": "console.flag/model",
+      "properties": {
+        "flag": "EXAMPLE",
+        "model": {
+          "group": "example.com",
+          "version": "v1",
+          "kind": "ExampleModel"
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Plugin Manifest Properties
+
+| Property | Description |
+| ----- | -------- |
+| `name: string` | Used as a unique identifier. Each plugin must have a unique name. |
+| `version: string` | The version of the plugin. Version must be parsable by node-semver, eg. “1.2.3” |
+| `pluginAPI: string` | Semver range of compatible OpenShift console versions. |
+| `displayName?: string` | User friendly display name. |
+| `description?: string` | The description of the plugin. |
+| `extensions: Extension[]` | List of extensions contributed by the plugin. |
+| `dependencies?: { [pluginName: string]: string }` | A plugin which depends on other plugins must explicitly define their dependencies. |
+
+(Types are expressed as TypeScript types.)
+
+##### Extension<P = JSON>
+
+| Property | Description |
+| ----- | -------- |
+| `id: string` | Unique identifier of this extension. |
+| `type: string` | The `type` identifies this extension of a particular extension type. |
+| `properties: P` | The extension properties JSON. |
+| `flags?: Partial<{ required: string[]; disallowed: string[]; }>` | Reference to the flags used to trigger the enablement of this extension. |
+
+An extension may contain code references, encoded as object literal `{
+$codeRef: string }`. The value of `$codeRef` is `moduleName.exportName` or
+`moduleName` (equivalent to `moduleName.default`). Console will resolve the
+code references when needed.
+
+### Lazy Loading
+
+Console uses
+[code splitting](https://webpack.js.org/guides/code-splitting/) to only load
+JavaScript code when needed. Likewise, dynamic plugin code will only load when
```

Comment:
it's not clear to me how the code is loaded/referenced.  The plugin config resource only points to the metadata.json path.

how does the console know which plugin endpoint to proxy to for a given JS module?  and how does it know the url path to the js module?

Reply:
(and when does the console backend refresh this content, in case the plugin is updated?)

Reply:
The plugin manifest has a `$codeRef` property for different extension points. The console uses that to load the JS module.

The backend doesn't load anything. It's merely a proxy. Console frontend loads the module when the extension is first needed. You would have to refresh the browser to see the update.